### DISCUSSION TITLE
go package getting removed from go.mod

### DISCUSSION THREAD

**Discussion by Himani-relan** [renovatebot/renovate#10196] [Request Help]
I ran renovate bot against a go code repository. 
One of the packages in go.mod is 
github.com/prometheus/client_golang v1.8.0
which needs to be upgraded to v1.10.0. 
Renovate is able to get the latest version of this package. I can see it in logs. As part of  PR updates and release notes also it says that this package has been upgraded. But in the actual go.mod file, it's removing the package. What could be the reason of removing this package from go.mod ?

Logs:

               {
                 "managerData": {"lineNumber": 15, "multiLine": true},
                 "depName": "github.com/prometheus/client_golang",
                 "depType": "require",
                 "currentValue": "v1.8.0",
                 "datasource": "go",
                 "depIndex": 7,
                 "updates": [
                   {
                     "newVersion": "v1.10.0",
                     "newValue": "v1.10.0",
                     "bucket": "latest",
                     "newMajor": 1,
                     "newMinor": 10,
                     "updateType": "minor",
                     "releaseTimestamp": "2021-03-18T12:11:37.000Z",
                     "branchName": "renovate/all"
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/prometheus/client_golang",
                 "currentVersion": "v1.8.0",
                 "isSingleVersion": true,
                 "fixedVersion": "v1.8.0"
               },

DEBUG: gomod.updateDependency: v1.10.0 (repository=xxx, branch=renovate/all)
DEBUG: Updating github.com/prometheus/client_golang in go.mod (repository=xxx, branch=renovate/all)

**Comment by rarkins** [comment]
Most likely is that "go get" removes it. The debug logs at time of commit will show the exact go commands which were run.

**Reply by Himani-relan** [reply]
```
DEBUG: gomod.updateDependency: v1.10.0 (repository=XXX, branch=renovate/all)
DEBUG: Updating github.com/prometheus/client_golang in go.mod (repository=XXX, branch=renovate/all)

DEBUG: gomod.updateArtifacts(go.mod) (repository=XXX, branch=renovate/all)
DEBUG: Using GOPATH: /go (repository=XXX, branch=renovate/all)
DEBUG: go get command included (repository=XXX, branch=renovate/all)
       "cmd": "go",
       "args": "get -d ./..."
DEBUG: go mod tidy command included (repository=XXX, branch=renovate/all)
       "cmd": "go",
       "args": "mod tidy"
DEBUG: additional go mod tidy command included (repository=XXX, branch=renovate/all)
       "cmd": "go",
       "args": "mod tidy"
DEBUG: Executing command (repository=XXX, branch=renovate/all)
       "command": "go get -d ./..."
DEBUG: exec completed (repository=XXX, branch=renovate/all)
       "cmd": "go get -d ./...",
       "durationMs": 74082,
       "stdout": "",
       "stderr": "go: downloading github.com/gorilla/mux v1.8.0\ngo: downloading golang.org/x/text v0.3.4\ngo: downloading golang.org/x/sync v0.0.0-20201207232520-09787c993a3a\ngo: downloading github.com/google/uuid v1.1.1\ngo: downloading golang.org/x/net v0.0.0-20201110031124-69a78807bb2b\ngo: downloading golang.org/x/sys v0.0.0-20210309074719-68d13333faf2\ngo: downloading github.com/go-logr/logr v0.3.0\ngo: downloading github.com/golang/protobuf v1.4.3\ngo: downloading google.golang.org/protobuf v1.25.0\ngo get: upgraded github.com/prometheus/common v0.15.0 => v0.18.0\ngo get: upgraded golang.org/x/sys v0.0.0-20210225134936-a50acf3fe073 => v0.0.0-20210309074719-68d13333faf2\n"
DEBUG: Executing command (repository=XXX, branch=renovate/all)
       "command": "go mod tidy"
DEBUG: exec completed (repository=XXX, branch=renovate/all)
       "cmd": "go mod tidy",
       "durationMs": 2318,
       "stdout": "",
       "stderr": "go: downloading github.com/pact-foundation/pact-go v1.5.2\ngo: downloading github.com/hashicorp/logutils v1.0.0\ngo: downloading github.com/onsi/ginkgo v1.14.1\ngo: downloading github.com/onsi/gomega v1.10.2\ngo: downloading github.com/google/go-cmp v0.5.4\ngo: downloading github.com/hashicorp/go-version v1.2.0\ngo: downloading github.com/nxadm/tail v1.4.4\n"
DEBUG: Executing command (repository=XXX, branch=renovate/all)
       "command": "go mod tidy"
DEBUG: exec completed (repository=XXX, branch=renovate/all)
       "cmd": "go mod tidy",
       "durationMs": 875,
       "stdout": "",
       "stderr": ""
DEBUG: Returning updated go.sum (repository=XXX, branch=renovate/all)
DEBUG: Found updated go.mod after go.sum update (repository=XXX, branch=renovate/all)
```

**Reply by Himani-relan** [reply]
These are the logs around that time. 
Why will "go get" remove it  ?

**Reply by Himani-relan** [reply]
Below is the configuration of latest renovate bot i am using:

```json
{
	"dryRun": false,
	"logLevel": "debug",
	"endpoint": "https://<git-url>/api/v3",
	"includeForks": false,
	"groupName": "all",
	"repositories": [
		XXXX
	],
	"baseBranches": [
		"master"
	],
	"dependencyDashboard": false,
	"separateMajorMinor": false,
	"automerge": false,
	"hostRules": [
		{
			"hostName": "github.com",
			"concurrentRequestLimit": 2
		}
	],
	"allowPostUpgradeCommandTemplating": true,
	"allowedPostUpgradeCommands": ["^jq .", "awk .", "mv ."],
	"trustLevel": "high",
	"rebaseWhen": "behind-base-branch",
}
```

Is there anything else i need to put in configuration ?

**Reply by rarkins** [reply]
Try running those same commands one by one locally to see which command removes the dependency. It could also be `go mod tidy`. If it's `go` doing it then I cannot tell you *why*

**Reply by viceice** [reply]
Migrate your config please:
* remove `"logLevel": "debug",` it's no longer supported, use `LOG_LEVEL=debug` environment.
* replace `"trustLevel": "high",`, it's no longer supported, use `allowCustomCrateRegistries` , `allowScripts` and `exposeAllEnv`.
* rename `hostName` => `matchhost`

**Reply by Himani-relan** [reply]
@rarkins : Yes it's go mod tidy which is removing these packages.
How can i disable this option ? I haven't explicitly defined to run gomodTidy in postUpdateOptions.

**Reply by rarkins** [reply]
Go mod tidy should otherwise only run if it's a major update, but according to your original description it's a minor update?

But doesn't the fact that go thinks the dependency is unused represent a problem with the project?

**Reply by panchn-n** [reply]
Thanks for the insight @rarkins . We do have a Makefile in our project and openapi generator commands that we run which results in additional go files. When renovate runs `go mod tidy`, those dependencies are cleaned up as those generated files are missing.
Is there a way to run commands like `make generate`, after `go mod tidy` is executed and before the dependencies are upgraded?

**Reply by rarkins** [reply]
Check out postUpgradeTasks (self-hosted Renovate only)

**Reply by panchn-n** [reply]
Thanks again @rarkins. It seems to have worked. I shall do more validation and close out this issue.

**Reply by Himani-relan** [reply]
Hey @rarkins , postUpgradeOptions didn't work for us. We are running renovate in a docker container. It's a scheduled job. So container spins up , get the latest config , runs and then exits.

Our use case is for make generate is:
- clone repo from git
- run make generate( it's spinning a docker container and doing the tasks). Adds file that's using prometheus packge.
- renovate does it's job to update dependencies.
Now can we clone the repo and run make generate by script instead renovate clones/fetches it from github repo. ?
Can we just clone it by ourselves, run make generate and mount that data to /tmp/renovate directory? Is it possible ?

If not, is there any other way to work around this issue ?

**Reply by viceice** [reply]
Nope, doesn't work. you probably need #8804 to be implemented.

**Reply by viceice** [reply]
Maybe a workaround is to use `postUpgradeTasks` with a custom renovate image. So you can make and go mod tidy with postUpgradeTasks.

Renovate customization:
```Dockerfile
FROM renovate/renovate:25.XX.XX

USER root

RUN install-apt make

USER 1000
```
`install-apt` is a custom [`containerbase/buildpack`](https://github.com/containerbase/buildpack)  command to install ubuntu packages. It will fetch packages feed and do cleanup after install automatically.

**Reply by Himani-relan** [reply]
Okay, i can try that option. One doubt : since make generate is going to generate files which we don't want to commit in github repo. So if i add make generate as part of postUpgradeTasks then those generated files will also be added in PR , right ?

**Reply by viceice** [reply]
Nope, there is an option which files from postupdate should be added to PR.

**Reply by Himani-relan** [reply]
Gotcha. Thanks @viceice , I will try that option.

**Reply by Himani-relan** [reply]
So i tried the option with below configuration:
        "allowedPostUpgradeCommands": ["^openapi-generator-cli .","go .","^test ."],
        "postUpgradeTasks": {
                "commands": ["test -f swagger-ui/openapi.json && openapi-generator-cli generate -i swagger-ui/openapi.json -g go-server -o src/ --package-name api ", "go mod tidy"],
                "fileFilters": ["go.mod"],
                "executionMode": "branch"
        },

It's able to generate the files. But still package which is part of generated files(github.com/prometheus/client_golang) is getting removed in go.mod file.
Here are the logs:

DEBUG: Executed post-upgrade task (repository=OPCC/container-launcher-service, branch=renovate/all)
       "dep": "APIDevTools/swagger-cli github.com/gorilla/mux github.com/pact-foundation/pact-go github.com/prometheus/client_golang html5lib ipaddress jq kubernetes paramiko pre-commit/pre-commit-hooks pyopenssl pytest pytest-dependency pytest-html pytest-json-report requests urllib3",
       "cmd": "go mod tidy",
       "stdout": "",
       "stderr": ""

DEBUG: Post-upgrade file saved (repository=OPCC/container-launcher-service, branch=renovate/all)
       "dep": "APIDevTools/swagger-cli github.com/gorilla/mux github.com/pact-foundation/pact-go github.com/prometheus/client_golang html5lib ipaddress jq kubernetes paramiko pre-commit/pre-commit-hooks pyopenssl pytest pytest-dependency pytest-html pytest-json-report requests urllib3",
       "file": "go.mod",
       "pattern": "go.mod"
